### DISCUSSION TITLE
Batch.use improvement

### DISCUSSION THREAD

**Discussion by hpn33** [libktx/ktx#167] [Ideas]
way dont use like this

```
inline fun <T : Batch> T.use(func: T.() -> Unit) {
    begin()
    func()
    end()
}
```
and on using
```
batch.use {
    draw(something, 0, 0)
// or
    player.draw(this)
}
```

**Comment by czyzby** [comment]
I'm not sure what you mean. [This utility is already available in `ktx-graphics`](https://github.com/libktx/ktx/blob/master/graphics/src/main/kotlin/ktx/graphics/graphics.kt#L27-L35).

**Comment by hpn33** [comment]
i say change that

```kotlin
//me
inline fun <T : Batch> T.use(func: T.() -> Unit) {
    begin()
    func() // <--
    end()
}

//you
inline fun <B : Batch> B.use(action: (B) -> Unit) {
  begin()
  action(this) // <--
  end()
}
```

on using
``` kotlin
// you
batch.use { it ->
   it.draw(somthing, 0, 0)
   it.draw(somthing, 0, 0)
   it.draw(somthing, 0, 0)

   player.draw(it)
   obj.draw(it)
}

//me
batch.use { it ->
  draw(somthing, 0, 0)
  draw(somthing, 0, 0)
  draw(somthing, 0, 0)

  player.draw(this)
  obj.draw(this)
}
```
I do not want to do a job twice


it`s just little better

**Comment by hpn33** [comment]
@czyzby

**Comment by czyzby** [comment]
That's deliberate, as you usually don't want to overshadow `this` when drawing stuff and you have the reference to the original batch you called `use` on. I'm usually very careful about using `T.()` extension methods, as they make it hard to reason about the scope of the block. So, the expected usage looks more like this:

```kotlin
batch.use {
   batch.draw(something, 0, 0)
   batch.draw(something, 0, 0)
   batch.draw(something, 0, 0)

   player.draw(batch)
   obj.draw(batch)
}
```

Besides, [the original `use` method](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html) from stdlib has the same signature.
